storage/sqlite/src/lib.rs

```diff
@@ -0,0 +1,337 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+extern crate core;
+
+mod error;
+
+use rusqlite::{Connection, OpenFlags, OptionalExtension};
+use std::borrow::Cow;
+use std::cmp::max;
+use std::path::PathBuf;
+use std::sync::{Mutex, MutexGuard};
+use std::vec::IntoIter;
+
+use crate::error::process_sqlite_error;
+use storage_core::{
+    backend::{self, TransactionalRo, TransactionalRw},
+    info::DbDesc,
+    Data, DbIndex,
+};
+use utils::shallow_clone::ShallowClone;
+use utils::sync::Arc;
+
+/// Return the table name that corresponds to a specific index
+/// TODO this is unoptimised and the returned values should be cached
+fn db_table_name(idx: usize) -> String {
+    format!("db_{}", idx)
+}
+
+/// Identifiers of the list of databases (key-value maps)
+#[derive(Eq, PartialEq, Debug, Clone)]
+struct DbList(Vec<()>);
+
+impl std::ops::Index<DbIndex> for DbList {
+    type Output = ();
+
+    fn index(&self, index: DbIndex) -> &Self::Output {
+        &self.0[index.get()]
+    }
+}
+
+/// Sqlite iterator over entries with given key prefix
+pub struct PrefixIter {
+    /// Underlying iterator
+    iter: IntoIter<(Vec<u8>, Vec<u8>)>,
+
+    /// Prefix to iterate over
+    prefix: Data,
+}
+
+impl PrefixIter {
+    fn new(iter: IntoIter<(Vec<u8>, Vec<u8>)>, prefix: Data) -> Self {
+        PrefixIter { iter, prefix }
+    }
+}
+
+impl Iterator for PrefixIter {
+    type Item = (Data, Data);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let kv = self.iter.next()?;
+        utils::ensure!(kv.0.starts_with(&self.prefix));
+        Some(kv)
+    }
+}
+
+pub struct DbTx<'m> {
+    connection: MutexGuard<'m, Connection>,
+}
+
+impl<'m> DbTx<'m> {
+    pub fn start_transaction(connection: MutexGuard<'m, Connection>) -> storage_core::Result<Self> {
+        let tx = DbTx { connection };
+        tx.connection.execute("BEGIN TRANSACTION", ()).map_err(process_sqlite_error)?;
+        Ok(tx)
+    }
+
+    pub fn commit_transaction(&self) -> storage_core::Result<()> {
+        let _res = self
+            .connection
+            .execute("COMMIT TRANSACTION", ())
+            .map_err(process_sqlite_error)?;
+        Ok(())
+    }
+}
+
+impl Drop for DbTx<'_> {
+    fn drop(&mut self) {
+        if self.connection.is_autocommit() {
+            return;
+        }
```

Comment:
Would the backend transaction mechanism behave correctly if autocommit is set to `true`? Maybe this should be an assertion rather than a check that silently handles that case.

Reply:
This is needed otherwise ROLLBACK TRANSACTION might fail

Reply:
I was thinking whether we should make sure at the beginning that autocommit is false, and have only an assertion here. Not preforming transaction rollback may leave the database in an unexpected state, so we should make effort to prevent it. But then again I am not sure what exactly the semantics of autocommit are, so maybe it's fine like this.